File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys


def dfs(deep, now):
    global result, end
    if deep == 5:
        result = 1
        end = True
        return

    for i in range(len(friend[now])):
        if end:
            return
        if not visit[friend[now][i]]:
            visit[friend[now][i]] = True
            dfs(deep+1, friend[now][i])
            visit[friend[now][i]] = False


end = False
result = 0
n, m = map(int, sys.stdin.readline().split())

friend = [[] for i in range(n)]
visit = [False] * n
for i in range(m):
    a, b = map(int, sys.stdin.readline().split())
    friend[a].append(b)
    friend[b].append(a)

for i in range(n):
    if result == 0:
        visit[i] = True
        dfs(1, i)
        visit[i] = False
print(result)

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static int n, m;
	static ArrayList<ArrayList<Integer>> friend;
	static int answer = 0;
	static boolean[] visit;
	static void dfs(int depth, int now) {
		if(depth == 4) {
			answer = 1;
			return;
		}
		for(int i = 0; i<friend.get(now).size(); i++) {
			if(!visit[friend.get(now).get(i)]) {
				visit[friend.get(now).get(i)] = true;
				dfs(depth+1, friend.get(now).get(i));
				visit[friend.get(now).get(i)] = false;
			}
		}
	}
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		
		friend = new ArrayList<>();
		visit = new boolean[n];
		for(int i = 0; i<n; i++) {
			friend.add(new ArrayList<>());
		}
		for(int i = 0; i<m; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			friend.get(a).add(b);
			friend.get(b).add(a);
		}
		for(int i = 0; i<n; i++) {
			if(answer == 1) break;
			visit[i] = true;
			dfs(0, i);
			visit[i] = false;
		}
		System.out.println(answer);
	}
}